helix-term/src/commands/typed.rs

```diff
@@ -2460,3 +2472,58 @@ pub(super) fn command_mode(cx: &mut Context) {
     prompt.recalculate_completion(cx.editor);
     cx.push_layer(Box::new(prompt));
 }
+
+fn expand_args<'a>(editor: &mut Editor, args: &'a str) -> Cow<'a, str> {
+    let reg = Regex::new(r"%(\w+)\s*\{(.*)").unwrap();
+    reg.replace(args, |caps: &regex::Captures| {
+        let remaining = &caps[2];
+        let end = find_first_open_right_braces(remaining);
+        let exp = expand_args(editor, &remaining[..end]);
+        let next = expand_args(editor, remaining.get(end + 1..).unwrap_or(""));
+        let (_view, doc) = current_ref!(editor);
+        format!(
+            "{} {}",
+            match &caps[1] {
+                "val" => match exp.trim() {
+                    "filename" => doc.path().and_then(|p| p.to_str()).unwrap_or("").to_owned(),
+                    "dirname" => doc
+                        .path()
+                        .and_then(|p| p.parent())
+                        .and_then(|p| p.to_str())
+                        .unwrap_or("")
+                        .to_owned(),
+                    _ => "".into(),
+                },
+                "sh" => {
+                    let shell = &editor.config().shell;
+                    if let Ok((output, _)) = shell_impl(shell, &exp, None) {
+                        output.trim().into()
+                    } else {
+                        "".into()
+                    }
+                }
+                _ => "".into(),
+            },
+            next
+        )
+    })
+}
+
+fn find_first_open_right_braces(str: &str) -> usize {
+    let mut left_count = 1;
+    for (i, &b) in str.as_bytes().iter().enumerate() {
+        match char::from_u32(b as u32) {
```

Comment:
Shouldn't we iterate over char directly instead? I think unicode might crash in this case when we are reading bytes within multiple bytes char.

Reply:
the `b` is a byte so that this won't crash.As we only use ascii chars in our patterns below, I didn't treat them as Unicode char, this shall work until we introduce other  Unicode pattern beside `%val` etc.

Reply:
The returned index will always be in bounds (the enumerator and len are both byte based).  Trying something like the following doesn't cause any issues
`sh echo %val{𒀀𒀀𒀀𒀀𒀀𒀀𒀀}` // 4 byte char f0928080

I think this should remain as is.

However, downstream I am seeing something off in shellwords (some multi-byte characters  are lost in the generated argument list, I am investigating it now but it is unrelated to this PR)